theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-contribution.ts

```diff
@@ -43,6 +82,32 @@ export class TraceViewerContribution extends WidgetOpenHandler<TraceViewerWidget
         }
     }
 
+    public async open(traceURI: URI, traceUUID?: TraceViewerWidgetOpenerOptions): Promise<TraceViewerWidget> {
+        try {
+            const healthResponse = await this.tspClient.checkHealth();
+            if ((healthResponse as TspClientResponse<HealthStatus>).getModel()?.status === 'UP') {
+                const widget = super.open(traceURI, traceUUID);
+                return widget;
+            }
+        }
+        catch (e) {
+            this.messageService.showProgress({
+                text: ''
+            }).then(async progress => {
+                    progress.report({ message: 'Launching trace server... ', work: { done: 10, total: 100 } });
+                    const resolve = await this.traceServerConfigService.startTraceServer(this.path, this.port);
+                    if (resolve === 'success') {
+                        progress.report({ message: `Trace server started on port: ${this.port}.`, work: { done: 100, total: 100 } });
+                        progress.cancel();
+                        const widget = super.open(traceURI, traceUUID);
+                        return widget;
+                    }
```

Comment:
in the case the start of the server is not successful, the still need to call progress.cancel() or done() otherwise the progress dialog stays open. Moreover, since the progress dialog is by default not cancelable, the dialog will stay open forever and the user can't close the dialog. I had this problem.

This also raises the question, if the progress monitor should be cancelable. At least the dialog should be closable. 

![image](https://user-images.githubusercontent.com/413424/111358760-15c34300-8661-11eb-9e6f-43389b5f458d.png)

There was also no indication that the server start failed (maybe a error dialog or status message?). I only saw it in the debug console:
 ERROR Request startTraceServer failed with error: spawn ./trace-compass-server/tracecompass-server ENOENT Error: spawn ./trace-compass-server/tracecompass-server ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:267:19)
    at onErrorNT (internal/child_process.js:469:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)

Reply:
Added progress.cancel() and toast messages in case there is an error starting the server.

The idea was not to open the dialog box if the server isn't started successfully. Yes, I have also noticed that the gethealth() API call takes some time to respond (especially when the trace server is down). To mitigate that I have added a spinner that says "Connecting to trace server" as an interim solution for now. In the future, I think we may need a better/faster solution to determine the state of the trace server.